### PR TITLE
Add revert changes button in local unit form

### PR DESCRIPTION
**Pull request by shreeyash07**
## Addresses:
- #1500 

## Depends on:
- https://github.com/IFRCGo/go-web-app/pull/1584

## Changes
- Add revert changes button in local unit form
- Add revert changes request
- Add revert changes modal form

## This PR doesn't introduce:
- [x] typos
- [x] conflict markers
- [x] unwanted comments
- [x] temporary files, auto-generated files or secret keys
- [x] `console.log` meant for debugging
- [x] codegen errors

### CODE REVIEW

**Review by barshathakuri — APPROVED** [review]

File: app/src/views/CountryNsOverviewContextAndStructure/NationalSocietyLocalUnits/LocalUnitsFormModal/LocalUnitsForm/i18n.json [hunk 1]
```diff
@@ -3,7 +3,9 @@
     "strings": {
         "lastUpdateLabel": "Last update: {modifiedAt} by {modifiedBy}",
         "successMessage": "Local unit added successfully!",
+        "revertChangesSuccessMessage": "Local unit reverted successfully!",
         "failedMessage": "Failed to add local unit",
+        "revertChangesfailedMessage": "Failed to revert changes in local unit",
```
Comment (tnagorra) [inline app/src/views/CountryNsOverviewContextAndStructure/NationalSocietyLocalUnits/LocalUnitsFormModal/LocalUnitsForm/i18n.json]:
```suggestion
        "revertChangesFailedMessage": "Failed to revert changes in local unit",
```

File: app/src/views/CountryNsOverviewContextAndStructure/NationalSocietyLocalUnits/LocalUnitsFormModal/LocalUnitsForm/index.tsx [hunk 2]
```diff
@@ -464,6 +487,44 @@ function LocalUnitsForm(props: Props) {
...
+            } = response;
+
+            alert.show(
+                strings.revertChangesfailedMessage,
+                {
+                    variant: 'danger',
+                    description: messageForNotification,
+                    debugMessage,
+                },
+            );
```
Comment (tnagorra) [inline app/src/views/CountryNsOverviewContextAndStructure/NationalSocietyLocalUnits/LocalUnitsFormModal/LocalUnitsForm/index.tsx]: Why not set error to the form?

File: app/src/views/CountryNsOverviewContextAndStructure/NationalSocietyLocalUnits/LocalUnitsFormModal/LocalUnitsForm/index.tsx [hunk 3]
```diff
@@ -347,6 +368,8 @@ function LocalUnitsForm(props: Props) {
         },
     });
 
+    const readOnly = readOnlyFromProps || localUnitDetailsResponse?.is_locked;
```
Comment (tnagorra) [inline app/src/views/CountryNsOverviewContextAndStructure/NationalSocietyLocalUnits/LocalUnitsFormModal/LocalUnitsForm/index.tsx]: Also check when localUnitDetailsResponse is not defined

File: app/src/views/CountryNsOverviewContextAndStructure/NationalSocietyLocalUnits/LocalUnitsFormModal/LocalUnitsForm/index.tsx [hunk 4]
```diff
@@ -509,7 +571,9 @@ function LocalUnitsForm(props: Props) {
 
     return (
         <div className={styles.localUnitsForm}>
-            {readOnly && isDefined(actionsContainerRef.current) && (
+            {!localUnitDetailsResponse?.is_locked
```
Comment (tnagorra) [inline app/src/views/CountryNsOverviewContextAndStructure/NationalSocietyLocalUnits/LocalUnitsFormModal/LocalUnitsForm/index.tsx]: Also check when localUnitDetailsResponse is not defined

File: app/src/views/CountryNsOverviewContextAndStructure/NationalSocietyLocalUnits/LocalUnitsFormModal/LocalUnitsForm/index.tsx [hunk 5]
```diff
@@ -509,7 +571,9 @@ function LocalUnitsForm(props: Props) {
...
+                && readOnlyFromProps
+                && isDefined(actionsContainerRef.current) && (
```
Comment (tnagorra) [inline app/src/views/CountryNsOverviewContextAndStructure/NationalSocietyLocalUnits/LocalUnitsFormModal/LocalUnitsForm/index.tsx:645]: It's not correct to check for ref on render. Change in ref will not be updated on render.

**Review by tnagorra — CHANGES_REQUESTED** [review]

File: app/src/views/CountryNsOverviewContextAndStructure/NationalSocietyLocalUnits/LocalUnitsFormModal/LocalUnitsForm/index.tsx (hunk 5 above)
Comment (shreeyash07) [reply app/src/views/CountryNsOverviewContextAndStructure/NationalSocietyLocalUnits/LocalUnitsFormModal/LocalUnitsForm/index.tsx:645]: Please ask @frozenhelium about this logic, i have not made any changes regarding check for ref on render.

File: app/src/views/CountryNsOverviewContextAndStructure/NationalSocietyLocalUnits/LocalUnitsFormModal/LocalUnitsForm/index.tsx [hunk 6]
```diff
@@ -509,7 +574,10 @@ function LocalUnitsForm(props: Props) {
 
     return (
         <div className={styles.localUnitsForm}>
-            {readOnly && isDefined(actionsContainerRef.current) && (
+            {!localUnitDetailsResponse?.is_locked
+                && readOnlyFromProps
+                && isNotDefined(localUnitDetailsResponse)
```
Comment (tnagorra) [inline app/src/views/CountryNsOverviewContextAndStructure/NationalSocietyLocalUnits/LocalUnitsFormModal/LocalUnitsForm/index.tsx]: Shouldn't this be `isDefined(localUnitDetailsResponse)` ?
Also, this check should come before `!localUnitDetailsResponse?.is_locked`

File: app/src/views/CountryNsOverviewContextAndStructure/NationalSocietyLocalUnits/LocalUnitsFormModal/LocalUnitsForm/index.tsx [hunk 7]
```diff
@@ -592,6 +660,15 @@ function LocalUnitsForm(props: Props) {
                                 && (environment !== 'production')
                                 && (
                                     <div className={styles.actions}>
+                                        {localUnitDetailsResponse?.is_locked && (
```
Comment (tnagorra) [inline app/src/views/CountryNsOverviewContextAndStructure/NationalSocietyLocalUnits/LocalUnitsFormModal/LocalUnitsForm/index.tsx]: Let's drop optional chaining.

Also, shouldn't the logic be inverted? Do we show "revert" button if the localUnit is locked?

**Review by tnagorra — CHANGES_REQUESTED** [review]

File: app/src/views/CountryNsOverviewContextAndStructure/NationalSocietyLocalUnits/LocalUnitsFormModal/LocalUnitsForm/index.tsx (hunk 7 above)
Comment (shreeyash07) [reply app/src/views/CountryNsOverviewContextAndStructure/NationalSocietyLocalUnits/LocalUnitsFormModal/LocalUnitsForm/index.tsx]: yes, we do show revert button if localunit is locked

**Review by samshara — APPROVED** [review]

**Review by samshara — APPROVED** [review]

**Pull request merged**